critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java

```diff
@@ -241,41 +244,59 @@ public void run() {
      * @throws InterruptedException
      * @throws WriteException
      */
-    private StorageMetadata syncArtifact(Iterator<URL> urlIterator, URI artifact, StorageAdapter sa)
-        throws StorageEngageException, InterruptedException, WriteException {
+    private StorageMetadata syncArtifact(Iterator<URL> urlIterator, URI artifact, StorageAdapter sa, ArtifactDAO artDao)
+        throws StorageEngageException, InterruptedException, WriteException, IllegalArgumentException {
 
         StorageMetadata storageMeta = null;
 
-        while (urlIterator.hasNext()) {
-            URL u = urlIterator.next();
-            log.debug("trying " + u);
-
-            try {
-                ByteArrayOutputStream dest = new ByteArrayOutputStream();
-                HttpGet get = new HttpGet(u, dest);
-                get.prepare();
-
-                NewArtifact a = new NewArtifact(artifact);
-                storageMeta = sa.put(a, get.getInputStream());
-                log.debug("storage meta returned: " + storageMeta.getStorageLocation());
-                return storageMeta;
-
-            } catch (WriteException wre) {
-                // IOException will capture this if not explicitly caught
-                log.info("write error for storage adapter put: " + wre.getMessage());
-                throw wre;
-            } catch (TransientException | IOException te) {
-                // ReadException will be caught under the IOException
-                // could be prepare or put throwing this error
-                // will move to next url
-                log.info("transient exception." + te.getMessage());
-
-            } catch (ResourceNotFoundException | ResourceAlreadyExistsException  badURL) {
-                log.info("removing URL " + u + " from list: " + badURL.getMessage());
-                urlIterator.remove();
-
+        if (urlIterator.hasNext()) {
+            while (urlIterator.hasNext()) {
+                URL u = urlIterator.next();
+                log.debug("trying " + u);
+
+                try {
+                    ByteArrayOutputStream dest = new ByteArrayOutputStream();
+                    HttpGet get = new HttpGet(u, dest);
+                    get.prepare();
+
+                    // Check content checksum and length to verify this
+                    // artifact can be synced currently
+                    Artifact curArtifact = artDao.get(artifact);
```

Comment:
It is a good thing to retrieve the Artifact from the database here because in principle it's state could have changed since the iterator returned the row and a FileSyncJob was created. It wasn't specified in the task, but at this point there should be a check that the artifact is still a valid thing to sync: if it now has a storageLocation, the fileSyncJob should stop/return with no further action.